scipy/stats/boost/tests/test_boost_stats.py

```diff
@@ -0,0 +1,77 @@
+'''Boost stats tests.'''
+
+import numpy as np
+from scipy.stats import (
+    bernoulli as scipy_bernoulli,
+    beta as scipy_beta,
+    nbinom as scipy_nbinom,
+    binom as scipy_binom,
+    ncx2 as scipy_ncx2,
+)
+from scipy.stats.boost import (
+    bernoulli as boost_bernoulli,
+    beta as boost_beta,
+    nbinom as boost_nbinom,
+    binom as boost_binom,
+    ncx2 as boost_ncx2,
+)
+
+
+def test_issue_11026():
+    p = 0.5
+    assert boost_bernoulli(p).median() == 0
+    assert boost_bernoulli(p).interval(1) == scipy_bernoulli(p).interval(1)
+    assert boost_bernoulli(p).interval(0.999) == scipy_bernoulli(p).interval(0.999)
+    assert boost_bernoulli(p).interval(1-1e-14) == scipy_bernoulli(p).interval(1-1e-14)
+
+def test_issue_12635():
+    p = 0.9999999999997369
+    a, b = 76.0, 66334470.0
+    tol = 1e-2
+    assert not scipy_beta.ppf(p, a-1, b) - scipy_beta.ppf(p, a, b) < tol
+    assert boost_beta.ppf(p, a-1, b) - boost_beta.ppf(p, a, b) < tol
+
+def test_issue_12794():
+    count_list = [10, 100, 1000]
+    p = 1e-11
+    sci = [scipy_beta.isf(p, count + 1, 100000 - count) for count in count_list]
+    boo = [boost_beta.isf(p, count + 1, 100000 - count) for count in count_list]
+    assert not all(np.diff(sci) > 0)
+    assert all(np.diff(boo) > 0)
+
+def test_issue_10317():
+    alpha, n, p = 0.9, 10, 1
+    assert scipy_nbinom.interval(alpha=alpha, n=n, p=p) != (0, 0)
+    assert boost_nbinom.interval(alpha=alpha, n=n, p=p) == (0, 0)
+
+def test_issue_11134():
+    alpha, n, p = 0.95, 10, 0
+    assert scipy_binom.interval(alpha=alpha, n=n, p=p) != (0, 0)
+    assert boost_binom.interval(alpha=alpha, n=n, p=p) == (0, 0)
+
+def test_issue_11777():
+    df, nc = 6700, 5300
+    n = 100
+    s = scipy_ncx2(df, nc)
+    b = boost_ncx2(df, nc)
+    assert not all(s.pdf(np.linspace(s.ppf(0.001), s.ppf(0.999), num=n)) > 0)
+    assert all(b.pdf(np.linspace(b.ppf(0.001), b.ppf(0.999), num=n)) > 0)
```

Comment:
Let's make this a stronger test, like the ones above, by comparing with known values.

Reply:
Now comparing against Gaussian approximation as mentioned in the issue.

I merged master and added `boost.ncx2` to all the tests in `stats/tests/test_distributions.py`.  The only failing continues to be `test_ncx2_tails_pdf[dist1]`:
```
>           assert_allclose(dist.logpdf(10000, 3, 12), -4662.444377524883)
...
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           x and y -inf location mismatch:
E            x: array(-inf)
E            y: array(-4662.444378)
```

Is `-4662.444378` actually the answer or is -inf acceptable here?

Reply:
Yeah, sounds like that should be reported.

Reply:
So I looked a little more and this is actually a precision issue.  For `double`s, Boost gives `ncx2(3, 12).pdf(10000)` as `0`, but with extended precison `long double`, it gives the correct `1.33701e-2025`, the log of which is `-4662.44...`.  That is a really small number and I wonder if it's actually useful to report all those decimals.  I think Boost's `double` ncx2 is reasonable in reporting 0 and users can ask for more precision by giving an `np.longdouble` type if needed.  Thoughts?

Reply:
For now, I think the easiest thing is to just keep the SciPy implementation of `ncx2` for this first PR. It's great to know that Boost supports extended precision, though, and we can mention this as a selling point.

So for all of these distributions we can get extended precision just by passing in a `np.longdouble`?

Reply:
I'm actually not sure it's setup to do that right now -- it's building ufuncs for `double` and `float` only because I remember having some failing tests for `long double`.  I can try again and see what those issues were

Reply:
Enabling the `long double` in the ufuncs leads to a lot of errors that looks like this:
```
E   TypeError: Cannot cast array data from dtype('float128') to dtype('float64') according to the rule 'safe'
```

I can keep looking into this if we think it will be worthwhile to have the extended precision

Reply:
It can wait if it's not easy. It would be great, but it's not part of the original scope.

Reply:
FYI: looks like previously there hasn't been a lot of enthusiasm about supporting `long double`, but that might have changed (see e.g. https://github.com/scipy/scipy/issues/10109)

Reply:
Looking into this actually helped find a few bugs in the cython code generation templates.

What I've learned:
- There appears to be an undocumented lookup order for ufunc types (smaller types should come first, larger last).  When I had `long double` declared first, it would always use the `long double` implementation.  When it was declared last, it's only used if the inputs are `long double`, otherwise it uses a smaller type
- `continuous_rv.pdf()` will return at least an `np.float64`.  It will up-cast an `np.float32` to an `np.float64` before returning.  Calling the boost implementation more or less directly (`scipy.stats.beta._pdf`) will return the appropriate type:
```pythonimport numpy as np
from scipy.stats import beta
for func, lbl in ((beta.pdf, 'pdf'), (beta._pdf, '_pdf')):
    print(f'{lbl}:\t', end='')
    for t in (np.float32, np.float64, np.float128):
        print(func(t(.6), t(.1), t(.2)).dtype, end=' ')
    print()
# output:
# pdf:	float64 float64 float128 
# _pdf:	float32 float64 float128 
``` 

I'll go ahead and add `long double` support to the ufuncs because it doesn't cost us anything and produces no additional test errors.  Happy to remove if it becomes a sore spot

Reply:
As you probably saw in the linked issues, supporting `long double` can be a problem.  The meaning of `long double` is platform dependent.  I think on some platforms it is true quad precision, while on others it is only 80 bits or even just 64 bits (i.e. an alias for `double`).

Reply:
But if is working, leave it in!  Just be aware that it might not be as high a precision as it looks on some platforms.

Reply:
It's just 64 bits on my machine, unfortunately : (